### DISCUSSION TITLE
Random generation of 2 identical emails

### DISCUSSION THREAD

**Discussion by vtgn** [faker-js/faker#2456] [Q&A] [answered]
Hi!

I use faker to generate some random data in my tests, but sometimes it generates 2 identical emails in the same test.
Last time, it happened in a test generating 45 emails, and 2 of them were identical (Reagan8@gmail.com), leading to the fail of my test because they need to be unique (it uses faker.internet.email() without any argument)

What is the probability that it happens?
Does faker generate a limited number of different emails (with the version without any argument)? How many?
Is there a tool or a solution given by faker to ensure this unicity? If not, it could be a good idea to implement one. ;)

Thank you.

Regards.

**Comment by ST-DDT** [comment, marked as answer]
Which locale are you using?
Do you use no/a random or a fixed seed?

There is faker.helpers.unique

**Reply by vtgn** [reply]
Hello!
![image](https://github.com/faker-js/faker/assets/112889052/a39876ba-490f-4df9-95c6-593424080306)

I don't set anything, I just call faker.internet.email() directly.

I will see faker.helpers.unique.

Thank you.

**Comment by matthewmayer** [comment]
In general it's quite common to get duplicates. Faker pulls from a fairly small set of first names, last names, and random numbers, and then you have an example of the birthday paradox - https://en.wikipedia.org/wiki/Birthday_problem - even if the number of possibilities is quite large, a relatively small number of attempts is likely to cause a duplicate .

**Comment by vtgn** [comment]
I've seen faker.helpers.unique and faker.helpers.uniqueArray, and it seems useful to provide what I need.
The only problem is what will happen if the generative function given as argument generates a too limited number of unique elements: I suppose it could take a long time to get the expected number of unique elements, or worse could be stuck in an infinity loop. :/

Thank you @ST-DDT for your hint. ;)

Regards.
